jwql/tests/test_utils.py

```diff
@@ -225,7 +247,86 @@
       'source_id': 's00001',
       'detector': 'Unknown',
       'file_root': 'jw80600-o009_s00001_miri_f1130w',
-      'group_root': 'jw80600-o009_s00001_miri_f1130w'}),
+      'group_root': 'jw80600-o009_s00001_miri_f1130w',
+      'observation': '009'}),
+
+    # Test stage 3 files with epoch
+    ('jw80600-o009_s00001-epoch1_miri_f1130w_i2d.fits',
+        {'program_id': '80600',
+         'ac_id': 'o009',
+         'source_id': 's00001',
+         'epoch': '1',
+         'instrument': 'miri',
+         'optical_elements': 'f1130w',
+         'suffix': 'i2d',
+         'recognized_filename': True,
+         'filename_type': 'stage_3_source_id_epoch',
+         'file_root': 'jw80600-o009_s00001-epoch1_miri_f1130w',
+         'detector': 'Unknown',
+         'group_root': 'jw80600-o009_s00001-epoch1_miri_f1130w',
+         'observation': '009'}),
+
+    # Test stage 3 filename with c-based association and subarray
+    ('jw01189-c1009_t006_nircam_f300m-maskrnd-sub320a335r_i2d.fits',
+        {'program_id': '01189',
+         'ac_id': 'c1009',
+         'target_id': 't006',
+         'instrument': 'nircam',
+         'optical_elements': 'f300m-maskrnd',
+         'subarray': 'sub320a335r',
+         'suffix': 'i2d',
+         'recognized_filename': True,
+         'filename_type': 'stage_3_subarray_based',
+         'file_root': 'jw01189-c1009_t006_nircam_f300m-maskrnd-sub320a335r',
+         'detector': 'Unknown',
+         'group_root': 'jw01189-c1009_t006_nircam_f300m-maskrnd-sub320a335r',
+         'observation': '062'}),
+
+    # Test nirspec IFU and MOS stage 3 files
+    ('jw09230-o004_t001_nirspec_g395h_f290lp_s3d.fits x1d.fits',
+        {'program_id': '09230',
+         'ac_id': 'o004',
+         'target_id': 't001',
+         'instrument': 'nirspec',
+         'optical_elements': 'g395h_f290lp',
+         'suffix': 's3d',
+         'recognized_filename': True,
+         'filename_type': 'stage_3_nirspec_ifu_mos',
+         'file_root': 'jw09230-o004_t001_nirspec_g395h_f290lp',
+         'detector': 'Unknown',
+         'group_root': 'jw09230-o004_t001_nirspec_g395h_f290lp',
+         'observation': '004'}),
+
+    # Test stag3, nirspec source based
+    ('jw04735-o005_s000013418_nirspec_f100lp-g140h_x1d.fits',
```

Comment:
@bhilbert4 as always, thank you for writing tests. So each part of this filename has an associated "key/value" pair that the file parser can recognize and can construct a valid level 3 filename?

One question I have, I see this has a specific source ID, I am assuming this x1d is for a MSA or IFU exposure. Will the source id's always correspond to the same source when the pipeline generates all of these individual x1d images? Will source `s000013418` be the same source in this version of the pipeline, older and newer?